tests/adam_test.cpp

```diff
@@ -13,319 +13,279 @@
-  arma::mat coordinates = arma::mat("2.9; 1.9");
-  optimizer.Optimize(f, coordinates);
-
-  REQUIRE(coordinates(0) == Approx(3.0).margin(0.05));
-  REQUIRE(coordinates(1) == Approx(2.0).margin(0.05));
-}
-
-/**
- * Test the Adam optimizer on the Three-hump camel function.
- * This is to test the Three-hump camel function and not Adam.
- * This test will be removed later.
- */
-TEST_CASE("AdamThreeHumpCamelFunctionTest", "[AdamTest]")
-{
-  Adam optimizer(0.001, 2, 0.7, 0.999, 1e-8, 500000, 1e-9, false);
-  FunctionTest<ThreeHumpCamelFunction>(optimizer, 0.1, 0.01);
-}
-
-/**
- * Test that multiple runs of the Adam optimizer result in the exact same
- * result.  This specifically tests that the update policy is successfully
- * reset at the start of each optimization.
- */
-TEST_CASE("AdamResetPolicyTest", "[AdamTest]")
-{
-  Adam optimizer(0.5, 2, 0.7, 0.999, 1e-8, 5, 1e-3, false);
-  optimizer.ResetPolicy() = true;
-
-  SphereFunction f(2);
-
-  arma::mat coordinatesA = f.GetInitialPoint();
-  optimizer.Optimize(f, coordinatesA);
-
-  // A second run should produce the exact same results.
-  arma::mat coordinatesB = f.GetInitialPoint();
-  optimizer.Optimize(f, coordinatesB);
-
-  CheckMatrices(coordinatesA, coordinatesB);
-}
+ #include <ensmallen.hpp>
+ #include "catch.hpp"
+ #include "test_function_tools.hpp"
+
+ using namespace ens;
+ using namespace ens::test;
+
+ TEMPLATE_TEST_CASE("Adam_SphereFunction", "[Adam]", arma::mat, arma::fmat)
+ {
+   Adam optimizer(0.5, 2, 0.7, 0.999, 1e-8, 50000, 1e-3, false);
+   FunctionTest<SphereFunctionType<TestType, arma::Row<size_t>>, TestType>(
+       optimizer, 0.5, 0.2);
+ }
+
+ TEMPLATE_TEST_CASE("Adam_StyblinskiTangFunction", "[Adam]", arma::mat)
+ {
+   Adam optimizer(0.5, 2, 0.7, 0.999, 1e-8, 50000, 1e-3, false);
+   FunctionTest<StyblinskiTangFunction<TestType, arma::Row<size_t>>, TestType>(
+       optimizer, 0.5, 0.1);
+ }
+
+ TEMPLATE_TEST_CASE("Adam_McCormickFunctionFunction", "[Adam]", arma::mat)
+ {
+   Adam optimizer(0.5, 1, 0.7, 0.999, 1e-8, 50000, 1e-5, false);
+   FunctionTest<McCormickFunction, TestType>(optimizer, 0.5, 0.1);
+ }
+
+ TEMPLATE_TEST_CASE("Adam_MatyasFunctionFunction", "[Adam]", arma::mat)
+ {
+   Adam optimizer(0.5, 1, 0.7, 0.999, 1e-8, 50000, 1e-5, false);
+   FunctionTest<MatyasFunction, TestType>(optimizer, 0.1, 0.01);
+ }
+
+ TEMPLATE_TEST_CASE("Adam_EasomFunction", "[Adam]", arma::mat)
+ {
+   Adam optimizer(0.2, 1, 0.7, 0.999, 1e-8, 50000, 1e-5, false);
+   FunctionTest<EasomFunction, TestType>(optimizer, 1.5, 0.01);
+ }
+
+ TEMPLATE_TEST_CASE("Adam_BoothFunction", "[Adam]", arma::mat)
+ {
+   Adam optimizer(1e-1, 1, 0.7, 0.999, 1e-8, 50000, 1e-9, true);
+   FunctionTest<BoothFunction, TestType>(optimizer);
+ }
+
+ TEMPLATE_TEST_CASE("AMSGrad_SphereFunction", "[Adam]", arma::fmat)
+ {
+   AMSGrad optimizer(1e-3, 1, 0.9, 0.999, 1e-8, 50000, 1e-11, true);
+   FunctionTest<SphereFunctionType<TestType, arma::Row<size_t>>, TestType>(
+       optimizer, 0.5, 0.1);
+ }
```

Comment:
Should this and some of the other optimizer tests be in this file, or in another file?  I think they were originally in other files.